### PR TITLE
Fix missing symlink explorer decoration if file is opened from quickpick

### LINKED ISSUE #147628: Missing symlink explorer decoration if file is opened from quickpick
- VS Code Version: stable and from sources

Steps to Reproduce:

1. Clone this repo https://github.com/gitpod-io/website and open in vscode
2. Open `src/routes/docs/troubleshooting.md` using quick pick
3. :bug: Go to explorer view and look for `src/routes/docs` (it's a symlink to `gitpod/docs`) folder, it has no symlink decoration

### PR DESCRIPTION
**Pull request by jeanp413**
This PR fixes #147628

### CODE REVIEW

File: src/vs/workbench/services/decorations/browser/decorationsService.ts [hunk 1]
```diff
@@ -398,8 +398,8 @@ export class DecorationsService implements IDecorationsService {
 
 	private _keepItem(map: DecorationEntry, provider: IDecorationsProvider, uri: URI, data: IDecorationData | undefined): IDecorationData | null {
 		const deco = data ? data : null;
-		const old = map.set(provider, deco);
```
Comment (jeanp413) [inline src/vs/workbench/services/decorations/browser/decorationsService.ts:401]: `map.set` returns the map so `old` was always truthy

File: src/vs/workbench/contrib/files/browser/views/explorerDecorationsProvider.ts [hunk 2]
```diff
@@ -65,10 +65,10 @@ export class ExplorerDecorationsProvider implements IDecorationsProvider {
 		return this._onDidChange.event;
 	}
 
-	provideDecorations(resource: URI): IDecorationData | undefined {
+	async provideDecorations(resource: URI): Promise<IDecorationData | undefined> {
```
Comment (JacksonKearl) [inline src/vs/workbench/contrib/files/browser/views/explorerDecorationsProvider.ts:68]: I do not understand the purpose of this change. Why make it async? And throw?

File: src/vs/workbench/services/decorations/browser/decorationsService.ts (hunk 1 above)
Comment (JacksonKearl) [reply src/vs/workbench/services/decorations/browser/decorationsService.ts:401]: Seems like here decorations will not fire if `data` is undefined. Is that intended?

File: src/vs/workbench/contrib/files/browser/views/explorerDecorationsProvider.ts (hunk 2 above)
Comment (jeanp413) [reply src/vs/workbench/contrib/files/browser/views/explorerDecorationsProvider.ts:68]: If we return `undefined` then `decorationsService` will assume it has no decorations at all for this provider and it won't try to fetch them again, so I just make it a promise and reject it to trigger [this logic](https://github.com/microsoft/vscode/blob/ad5ef7ea85b8a622f73b96394507d98809e68861/src/vs/workbench/services/decorations/browser/decorationsService.ts#L389-L391) and decorations can be fetched when explorerItem is resolved

File: src/vs/workbench/services/decorations/browser/decorationsService.ts (hunk 1 above)
Comment (jeanp413) [reply src/vs/workbench/services/decorations/browser/decorationsService.ts:401]: Fixed it

**Comment by lramos15** [comment]
Bump @jrieken :)

File: src/vs/workbench/contrib/files/browser/views/explorerDecorationsProvider.ts (hunk 2 above)
Comment (jrieken) [reply src/vs/workbench/contrib/files/browser/views/explorerDecorationsProvider.ts:68]: Throwing an error is a little fishy... Can't this use the `onDidChange` event and trigger a refresh that way?

**Review by jrieken — COMMENTED** [review]
Changes in `decorationsService` look good, they fix an issue that was regressed after some perf-refactorings. Thanks for that 👏 

Unsure about the throw-control-flow. I'd say an event a better suited but I am no expert in the explorer domain

File: src/vs/workbench/contrib/files/browser/views/explorerDecorationsProvider.ts (hunk 2 above)
Comment (jeanp413) [reply src/vs/workbench/contrib/files/browser/views/explorerDecorationsProvider.ts:68]: > Can't this use the onDidChange event and trigger a refresh that way?

There isn't an event that fires when an [ExplorerItem is resolved](https://github.com/microsoft/vscode/blob/75a6ddc8626b831f01f91f9f60f27de3d5bb6f38/src/vs/workbench/contrib/files/browser/views/explorerViewer.ts#L96-L140), maybe we can add another emitter like `explorerRootErrorEmitter` but that doesn't look nice either. Any other ideas?

**Review by lramos15 — APPROVED** [review]

**Review by jrieken — APPROVED** [review]

**Pull request merged**